client-ui/src/app/model/operation/operation-property.models.ts

```diff
@@ -42,77 +42,61 @@ export namespace OperationProperty {
     json: ExtraPropertyValueJson[]
   }
 
-  export function decorateJsonProperty(propValue: ExtraPropertyValueJson): ExtraPropertyValueJson {
-    // make label human readable "camelCase" => "Camel Case"
-    const regex = new RegExp('([a-z])([A-Z])', 'g')
-    let label = propValue.label.replace(regex, '$1 $2')
-    label = label.substring(0, 1).toUpperCase() + label.substring(1)
-    return {
-      ...propValue,
-      label
-    }
-  }
-
-  export function decorateExpressionProperty(propValue: ExtraPropertyValue<Record<string, any>>,
-                                             attributesList: AttributeVM[]): ExtraPropertyValueExpression {
-    const jsonLikeProp = decorateJsonProperty(propValue)
-    return {
-      ...jsonLikeProp,
-      value: {
-        value: getText(propValue.value, attributesList),
-        rawValue: propValue.value
-      }
-    }
-  }
-
-  export function isExpresionProperty(propertyValue: Record<string, any>): boolean {
-    return propertyValue._typeHint && (propertyValue._typeHint as string).startsWith('expr.')
-  }
-
   export function parseExtraOptions(nativeProperties: NativeProperties, attributesList: AttributeVM[]): ExtraProperties {
-    const extraProperties: ExtraProperties = {
-      primitive: [],
-      expression: [],
-      json: []
-    }
-
-    return Object.keys(nativeProperties)
+    return Object.entries(nativeProperties)
+      .filter(([, v]) => v != null)
       .reduce(
-        (result, key) => {
-          const value = nativeProperties[key]
-
-          if (value === undefined || value === null) {
-            return result
+        (acc, [key, value]) => {
+          if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') {
+            const primitiveValue: ExtraPropertyValuePrimitive = {
+              label: key,
+              value: value
+            }
+            return {...acc, primitive: [...acc.primitive, primitiveValue]}
```

Comment:
you sohuld not be copying the whole object for just push another array  item here. 

acc.primitive.push(primitiveValue)

Reply:
I would agree if the algorithm were mutable. But in this case `reduce` function is used, which implies immutability. And modifying an accumulator from inside a reducer is quite a mistake.

Reply:
Why the algorithm should be immutable in that case? There is no need to use more memory and make code less readable here. 

What kind of mistake you are talking about? I cannot see any mistake by working with one object all the time inside a reduce function and just push new values to the array in the current case.

Reply:
We prefer functional code style for everything unless there are reasons to do opposite.

Reply:
> I cannot see any mistake by working with one object all the time inside a reduce function and just push new values to the array in the current case.

Of cause, technically it's not an issue. The compiler would be happy. The problem is laying in violation one of core FP principles - _breaking immutability_.  It is extremely important from the perspective or code reading and maintenance. In the above example by mutating `acc` instance from inside a reducer function you defeat the purpose of an accumulator, and effectively turn `reduce` function into `forEach`-ish. The resulting code however doesn't resemble neither of those, and raises too many "why"s to a reader. 

Sure, paradigms can be mixed in practical coding, but it should be done with care, in a cohesive manner, avoiding collisions. After all, programming paradigms exist to decrease the level of entropy in software development - their main purpose is to create restriction rather than giving new possibilities. So it's very important to understand and obey their core principles.

Reply:
I do not understand what we are talking about here. It does not break any principle as far we know what you are doing here. We are working with an object we have created by our own and which exists only in the current scope. It will not bring any unexpected behavior and any potential bugs, plus it solves a few problems: 
- it is better to read
- it is faster
- it uses less memory.

If you have a big problem with the mutating accumulator, then do not use reduce function here. As it is not a good use case in your case, then just work with forEach. But do not create a bunch of copy objects.

Reply:
Let's discuss it on CQC